feature: admissionControl filter (#2030)
zalando/skipper: docs/reference/filters.md

### admissionControl

Implements an admission control filter similar to [envoy admission control](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/admission_control_filter#admission-control).

The probability of rejection is calculated by the following equation:

$$ P_{reject} = ( { n_{total} - { n_{success} \over threshold } \over n_{total} + 1} )^{1 \over aggression } $$

Examples:

    admissionControl(metricSuffix, mode, d, windowSize, minRPS, successThreshold, maxRejectProbability, aggression)
    admissionControl("myapp", "active", "1s", 5, 10, 0.95, 0.9, 2.0)

Parameters:

* metric suffix (string)
* mode (enum)
* d (time.Duration)
* window size (int)
* min requests (int)
* success threshold (float64)
* max reject probability (float64)
* aggression (float64)

Metric suffix is the chosen suffix key to expose reject counter,
should be unique by filter instance

Mode has 3 different possible values and defaults to "inactive":

* "active" will reject traffic
* "inactive" will never reject traffic (default)
* "log" will not reject traffic, but log to debug filter settings

D the time duration of a single slot for required counters in our

> circular buffer of window size and default to 1s.

Reviewer: For leaky bucket I decided not to use any defaults to make users understand what they are doing and because there is no one-size-fits all config for load schedding/ratelimit IMO.

Author: I want to adjust defaults when we know there might be known good configuration for the bulk of services at best.
I don't think it's reasonable that the users that have no idea about the thing can judge if windowSize should be 5 or 50 and I want to make default a good choice and I tried to sort it such that simple things are in front and complicated things are last.
So it was a careful choice, that I would like to keep.

Author: Agreed on first without defaults and later if we observe known good values, we can add defaults

Window size is the size of the circular buffer. It is used to snapshot
counters to calculate total requests and number of success. It is
within $[1, 100]$ and defaults to 10.

Min requests is the minimum requests per $windowSize * d$ that have to pass this filter

> otherwise it will not reject traffic.

Reviewer: In the [envoy](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/admission_control_filter#overview) it is requests per second (rps), not per `$windowSize * d$`

Author: Yes we can do better than envoy

Author: we discussed and the purpose of having rps is for the user as input to us.

Success threshold sets the lowest request success rate at which the
filter will not reject requests. It is within $(0,1]$ and defaults to
0.95, which means an error rate of lower than 5% will not trigger
rejects.

Max reject probability sets the upper bound of reject probability. It
is within (0,1] and defaults to 0.95, which means if backend errors
with 100% it will only reject up to 95%.

aggression is used to dictate the rejection probability. The
calculation is done by $p = p^{1 \over aggression}$

> The aggression value is within $(0,\infty]$:

Reviewer: Aggression is similar to `fadeIn`'s `exponent`: `exponent = 1/aggression`.

Author: What is the suggestion?

Author: I changed it to exponent, I like this also more!

Author: Please check if the change makes sense to you.

Path("/expensive") -> clusterLeakyBucketRatelimit("user-${request.cookie.Authorization}", 1, "1s", 5, 2) -> ...
```

## shedder

> The basic idea of load shedding is to reduce errors by early stopping
> partial of the ingress requests that create too much load and serve
> the maximum throughput the system can do.

Reviewer 2: ```suggestion
The basic idea of load shedding is to reduce errors by early stopping
some of the ingress requests that create too much load and serving
the maximum throughput the system can process at a point in time.
```

There is a great talk by [Acacio Cruz from

> Google](https://www.youtube.com/watch?v=XNEIkivvaV4&feature=youtu.be)

Reviewer 2: ```suggestion
Google](https://www.youtube.com/watch?v=XNEIkivvaV4)
```

Author: thanks fixed

that explains the basic principles.

### admissionControl

Implements an admission control filter, that rejects traffic by
observed error rate and probability.

The probability of rejection is calculated by the following equation:

$$ P_{reject} = ( { n_{total} - { n_{success} \over threshold } \over n_{total} + 1} )^{ exponent } $$

Examples:

    admissionControl(metricSuffix, mode, d, windowSize, minRPS, successThreshold, maxRejectProbability, exponent)
    admissionControl("myapp", "active", "1s", 5, 10, 0.95, 0.9, 0.5)

Parameters:

* metric suffix (string)
* mode (enum)
* d (time.Duration)
* window size (int)

> * min requests (int)

Reviewer 2: minRPS carries a slightly different meaning (per second)

Author: yes we discussed it elsewhere and I changed this behavior.
